source/Attribute.cpp

```diff
@@ -0,0 +1,425 @@
+Attribute *Attribute::Parse(const string &token)
+{
+	if(!cached)
+		PrepareCache();
+	auto it = oldToNew.find(token);
+	if(it == oldToNew.end())
+		return nullptr;
+	return it->second;
+}
+
+
+
+// Creates a cache for faster data loading.
+void Attribute::PrepareCache()
+{
+	for(int effect = -1; effect < ATTRIBUTE_EFFECT_COUNT * 4; effect++)
+	{
+		for(int category = -1; category < ATTRIBUTE_CATEGORY_COUNT; category++)
+		{
+			for(int secondary = -1; secondary < (ATTRIBUTE_EFFECT_COUNT * 4) * (category == RESISTANCE); secondary++)
+			{
+				Attribute *a = new Attribute(static_cast<AttributeCategory>(category), static_cast<AttributeEffect>(effect),
+						static_cast<AttributeEffect>(secondary));
+				string s = a->CalculateLegacyName();
+				newToOld[category + 1][effect + 1][secondary + 1] = s;
+				if(oldToNew.find(s) == oldToNew.end())
+					oldToNew[s] = a;
+				else
+					delete a;
+			}
+		}
+	}
+	// ensure that effect names are always recognized
+	for(int effect = 0; effect < ATTRIBUTE_EFFECT_COUNT * 4; effect++)
+	{
+		string s = GetEffectName(static_cast<AttributeEffect>(effect));
+		if(oldToNew.find(s) == oldToNew.end())
+			oldToNew[s] = new Attribute(static_cast<AttributeCategory>(-1), static_cast<AttributeEffect>(effect),
+					static_cast<AttributeEffect>(-1));;
+	}
+	cached = true;
+}
+
+
+
+// Checks whether the specified combination is fully supported by the engine.
+bool Attribute::IsSupported() const
+{
+	if(effect == -1)
+		return false;
+	if(secondary != -1)
+	{
+		if(category != RESISTANCE)
+			return false;
+		if(effect >= ATTRIBUTE_EFFECT_COUNT || effect == PIERCING)
+			return false;
+		if(!Attribute(category, effect).IsSupported())
+			return false;
+		return secondary == ENERGY || secondary == HEAT || secondary == FUEL;
+	}
+	if(IsRelative())
+	{
+		if(IsMultiplier())
+			return false;
+		int effectType = effect % ATTRIBUTE_EFFECT_COUNT;
+		switch(category)
+		{
+			case DAMAGE:
+				if(effectType == DISABLED || effectType == MINABLE)
+					return true;
+			case FIRING:
+				if(effectType <= HULL || (effectType >= ENERGY && effectType <= HEAT))
+					return true;
+			default:
+				return false;
+		}
+	}
+	if(IsMultiplier())
+	{
+		if(category != SHIELD_GENERATION && category != HULL_REPAIR)
+			return false;
+		int effectType = effect % ATTRIBUTE_EFFECT_COUNT;
+		return effectType == ENERGY || effectType == HEAT || effectType == FUEL || effectType == category;
+	}
+	if(effect == THRUST && category == AFTERBURNING)
+		return true;
+	if(effect >= THRUST && effect <= CLOAK)
+		return static_cast<int>(effect) == static_cast<int>(category);
+	if(effect >= ENERGY && effect <= HEAT)
+		return (category != COOL && category != ACTIVE_COOL && category != RESISTANCE) ||
+				(effect == ENERGY && category == ACTIVE_COOL);
+	if(effect >= DISCHARGE && effect <= DISRUPTION)
+		return (category >= AFTERBURNING && category <= DAMAGE) || (category >= THRUSTING && category <= TURNING);
+	switch(effect)
+	{
+		case SHIELDS:
```

Comment:
this whole thing needs comment but why's this one empty

Reply:
It's a switch-case thing. Cases without break continue to the case below them.

Effectively, the code for SHIELDS and HULL is the same, so SHIELDS gets an empty case and executes the code below it.

Reply:
Also, this function will be massively simplified once we actually start using this system. But yeah, I'll add more comments here.